dm/pkg/binlog/pos_finder_test.go

```diff
@@ -0,0 +1,447 @@
+		c.Assert(err, IsNil)
+		c.Assert(location.Position, Equals, mysql.Position{"mysql-bin.000003", targetEventStart})
+		c.Assert(location.GTIDSetStr(), Equals, "")
+		c.Assert(posType, Equals, InRangeBinlogPos)
+	}
+}
+
+func (t *testPosFinderSuite) TestErrorCase(c *C) {
+	flavor := "mysql"
+	relayDir := c.MkDir()
+	beforeTime := time.Now()
+	tcctx := tcontext.NewContext(context.Background(), log.L())
+	{
+		finder := NewLocalBinlogPosFinder(tcctx, true, flavor, relayDir+"not-exist")
+		_, _, err := finder.FindByTs(beforeTime.Add(-time.Minute).Unix())
+		c.Assert(err.Error(), Matches, ".*no such file or directory.*")
+	}
+	{
+		finder := NewLocalBinlogPosFinder(tcctx, true, flavor, c.MkDir())
+		_, _, err := finder.FindByTs(beforeTime.Add(-time.Minute).Unix())
+		c.Assert(err.Error(), Matches, ".*cannot find binlog files.*")
+	}
+	{
+		file, err := os.Create(path.Join(relayDir, "mysql-bin.000001"))
+		c.Assert(err, IsNil)
+		file.Close()
+		finder := NewLocalBinlogPosFinder(tcctx, true, flavor, relayDir)
+		_, _, err = finder.FindByTs(beforeTime.Add(-time.Minute).Unix())
+		c.Assert(err.Error(), Matches, "EOF")
+	}
+}
+
+func (t *testPosFinderSuite) TestMySQL57GTID(c *C) {
+	flavor := "mysql"
+	relayDir := c.MkDir()
+	beforeTime := time.Now()
+	latestGTIDStr := "ffffffff-ffff-ffff-ffff-ffffffffffff:1"
+
+	generator, _ := event.NewGeneratorV2(flavor, "5.7.0", latestGTIDStr, true)
+
+	file1Events, data := genBinlogFile(generator, beforeTime, "mysql-bin.000002")
+	c.Assert(len(file1Events), Equals, 15)
+	_ = os.WriteFile(path.Join(relayDir, "mysql-bin.000001"), data, 0644)
+	file2Events, data := genBinlogFile(generator, beforeTime.Add(5*time.Second), "mysql-bin.000003")
+	c.Assert(len(file2Events), Equals, 15)
+	_ = os.WriteFile(path.Join(relayDir, "mysql-bin.000002"), data, 0644)
+	file3Events, data := genBinlogFile(generator, beforeTime.Add(10*time.Second), "mysql-bin.000004")
+	c.Assert(len(file3Events), Equals, 15)
+	_ = os.WriteFile(path.Join(relayDir, "mysql-bin.000003"), data, 0644)
+
+	tcctx := tcontext.NewContext(context.Background(), log.L())
+
+	{
+		finder := NewLocalBinlogPosFinder(tcctx, true, flavor, relayDir)
+		location, posType, err := finder.FindByTs(beforeTime.Add(-time.Minute).Unix())
+		c.Assert(err, IsNil)
+		c.Assert(location.Position, Equals, mysql.Position{"mysql-bin.000001", 4})
+		c.Assert(location.GTIDSetStr(), Equals, "ffffffff-ffff-ffff-ffff-ffffffffffff:1")
+		c.Assert(posType, Equals, BelowLowerBoundBinlogPos)
+	}
+	{
+		gtids := []string{
+			"ffffffff-ffff-ffff-ffff-ffffffffffff:1",
+			"ffffffff-ffff-ffff-ffff-ffffffffffff:1-2",
+			"ffffffff-ffff-ffff-ffff-ffffffffffff:1-3",
+		}
+		var targetEventStart uint32
+		var targetEvent *replication.BinlogEvent
+		cnt := 0
+		for _, ev := range file1Events {
+			if ev.Header.EventType == replication.GTID_EVENT {
+				targetEvent = ev
+
+				finder := NewLocalBinlogPosFinder(tcctx, true, flavor, relayDir)
```

Comment:
this is a light suggestion because most logic for local and remote are seem, we can do this test in next PR's it-test

seems all tests use local finder, do we need test remote finder? can use `MockStream`

Reply:
most logic is location neutral, either local or remote, so let `GetBinaryLogs` and `NewTCPReader` test themself.